snuba/datasets/errors_replacer.py

```diff
@@ -192,25 +192,37 @@ def set_project_exclude_groups(
     )
     p = redis_client.pipeline()
 
-    group_id_data: Mapping[str | bytes, bytes | float | int | str] = {
-        str(group_id): now for group_id in group_ids
-    }
+    group_id_data: MutableMapping[str | bytes, bytes | float | int | str] = {}
+    for group_id in group_ids:
+        group_id_data[str(group_id)] = now
+        if len(group_id_data) > settings.REPLACER_MAX_GROUP_IDS_TO_EXCLUDE:
+            break
+
     p.zadd(key, group_id_data)
-    # remove group id deletions that should have been merged by now
-    p.zremrangebyscore(key, -1, now - settings.REPLACER_KEY_TTL)
+    truncate_group_id_replacement_set(p, key, now)
     p.expire(key, int(settings.REPLACER_KEY_TTL))
 
     # store the replacement type data
     replacement_type_data: Mapping[str | bytes, bytes | float | int | str] = {
         replacement_type: now
     }
     p.zadd(type_key, replacement_type_data)
-    p.zremrangebyscore(type_key, -1, now - settings.REPLACER_KEY_TTL)
+    truncate_group_id_replacement_set(p, type_key, now)
     p.expire(type_key, int(settings.REPLACER_KEY_TTL))
 
     p.execute()
 
 
+def truncate_group_id_replacement_set(p: Any, key: str, now: float) -> None:
+    # remove group id deletions that should have been merged by now
+    p.zremrangebyscore(key, -1, now - settings.REPLACER_KEY_TTL)
+    # remove group id deletions that exceed the maximum number of deletions
+    # snuba's query processor will put in a query.
+    # Add +1 such that the query processor will still recognize that we
+    # exceeded the limit and fall back to FINAL.
+    p.zremrangebyrank(key, 0, -(settings.REPLACER_MAX_GROUP_IDS_TO_EXCLUDE + 1))
```

Comment:
Is my understanding correct?

Previous logic:
Remove all stale replacements. Relevant replacements still remain in Redis - no limit on the total number.

New logic:
Remove all stale replacements. Relevant replacements still remain but are limited to a maximum for this project.

`REPLACER_MAX_GROUP_IDS_TO_EXCLUDE` is the maximum number of group ids we add to a *single* query to exclude before we just set the query to final.

So there isn't any need in Redis to store more than that max (+ 1) since logically it will still be FINAL and the additional storage does not help?

Reply:
https://github.com/getsentry/snuba/blob/0a2f2bb79d9f56174bb495498b8bb363512ec668/snuba/datasets/storages/processors/replaced_groups.py#L100

the idea is that the query processor sees that there are strictly more group ids in redis than the maximum, and falls back to final

however, I just realized that this won't work at all, because the length check there is against the number of groups that would be excluded (based on whether there are group ids in the query). so it might actually misbehave

also I just remembered that filippo pointed out that we can just set the final flag for this project, so I'll do that now